Rework Warsong Gulch Battleground

BenDol opened the pull request:
## 🍰 Pullrequest
Reworked Warsong Gulch to be more efficient and resolve the strange sync issues.

Thanks to @Yatzii93 for testing and also providing the fix for the FillInitialWorldStates issue.

Deals with:
* The huge amount of code duplication
* Uses team index for all common data configs (FlagIds, MessageIds, MessageTypes, Sounds)
* Centralize flag actions to `PickUpFlagFromBase` & `GroundFlagInteraction`
* Properly checks the flags display id for the ground pick up actions

### Proof
Will compile a video of all the results, but it seems to be working flawlessly form all my testing.

### Issues
https://github.com/cmangos/issues/issues/1872
also addresses https://github.com/cmangos/mangos-tbc/pull/296

### How2Test
.debug bg
enter warsong

### Todo / Checklist
- [X] Internal Tests
- [x] Production Tests

Yatzii93 commented:
Will test very soon and report back.

killerwife commented:
Okay, let me say this outright. I will focus on your other PRs, and when you are done here, we will talk design.

BenDol commented:
There issues with the design I've used here?

killerwife commented:
Nope, or at least not as of right now. But until you are done, I'd rather focus on other things. Please do provide and extensive list of issues it addresses when you are done.

BenDol commented:
Ah okay I see, sure thing, I'm done now; extensively tested as much as I could by myself.

It mainly addresses the issues I posted but I just couldn't stand how poorly written it was with all the code duplication etc. So I just rewrote it to be a lot cleaner.

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -553,10 +474,12 @@ WorldSafeLocsEntry const* BattleGroundWS::GetClosestGraveYard(Player* player)
         return sWorldSafeLocsStore.LookupEntry(WS_GRAVEYARD_FLAGROOM_ALLIANCE);
     }
-    if (GetStatus() == STATUS_IN_PROGRESS)
+    else if (GetStatus() == STATUS_IN_PROGRESS)
```
Redundant else, please don't undo my work: https://github.com/cmangos/mangos-tbc/commit/ea4256273ba5b82e85fd88d9ba155d75b98ea5fb#diff-9f4e0b9638611b3f7a54b42f2c8fa91e

evil-at-wow replied:
I second that.

BenDol replied:
strange coding practices to remove an else  where even though logically it is still fine. The `else` from my experience also  makes it easier to read the code to understand the logical trail of the `if` statement. Removing them seems so pointless and deters away from the if statements logic trail. Doesn't make sense to remove them at all, unless I've missed something.

domgho replied:
1. Resharper and other code cleanup tools suggest it: https://www.jetbrains.com/help/resharper/RedundantIfElseBlock.html 2. It reduces nested code which you already noticed as you didn't add an else a few lines above/below either 3. It allows programmers to recognize this pattern quicker 4. It is currently consistent in the whole code base

BenDol replied:
Seems to be semantics at this point, I'll change it back for the sake of point `4` which is a valid point.

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -50,7 +50,7 @@ void BattleGroundWS::Update(uint32 diff)
             }
         }
-        if (m_FlagState[TEAM_INDEX_ALLIANCE] == BG_WS_FLAG_STATE_ON_GROUND)
+        else if (m_FlagState[TEAM_INDEX_ALLIANCE] == BG_WS_FLAG_STATE_ON_GROUND)
```
Good else

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -133,13 +131,13 @@ void BattleGroundWS::RespawnDroppedFlag(Team team)
 
-    PlaySoundToAll(BG_WS_SOUND_FLAGS_RESPAWNED);
+    SendMessageToAll(wsMessageIds[teamIdx][BG_WS_FLAG_ACTION_CAPTURED],
+        wsChatMessageTypes[teamIdx][BG_WS_FLAG_ACTION_CAPTURED]);
```
I’ll take this as an example for all your similar “optimizations”.. all you are doing is adding array lookups since the branch (maybe even two branches) still exists in GetTeamIndexByTeamId. Yes there is less code duplication but at the price of performance and i don’t think BG code needs to be touched often if it’s working

Yatzii93 replied:
Well to be fair, there was game breaking problems that needed to be addressed.

evil-at-wow replied:
No one said we can't touch the code if there are issues (game breaking or otherwise). I think he added "if it's working" as a condition for a reason :smile:

BenDol replied:
Which it wasn't & even this rework might have more issues. The point is that cleaner code makes for easier debugging, that was my whole reason to rework this class.

Yatzii93 replied:
I probably took it to literally :P I have a habit of doing that.

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -501,10 +424,8 @@ void BattleGroundWS::Reset()
-        RewardHonorToTeam(GetBonusHonorFromKill(m_HonorWinKills), ALLIANCE);
-    if (winner == HORDE)
-        RewardHonorToTeam(GetBonusHonorFromKill(m_HonorWinKills), HORDE);
+    RewardHonorToTeam(GetBonusHonorFromKill(m_HonorWinKills), winner);
```
What if there is no winner? (45m time limit)

BenDol replied:
Good point, thanks.

domgho commented:
Since I was the last one who cleaned up WS code I decided to review it for you. Personally I would not take the design changes and code duplication reduction because of inefficiency and I would ask you to only add the fixes that you apparently unearthed but I left the project some time ago to pursue new things. Hope it helps.

BenDol commented:
Nothing personal dude, was just not the most efficiently written class, all the logic and design was spot on (minus a few bugs). If you want to strain progress because you're unable to accept constructive criticism I you aren't the right person to be reviewing this PR. No harm was intended.

domgho commented:
I wanted to help you see it from a different perspective and gave you an example of why I think it’s slower and agreed that it reduces code duplication which I however personally value less than performance. You of course don’t need to take my feedback into account. No offense taken or intended

BenDol commented:
My only goal here is to help out this project. I don't care if this PR is merged or not. If its deemed to be too much of a performance hit for the sake of cleaner code, then I'll accept that from one of the core dev team. The reason I rewrote it in the first place was because debugging the issues was hell since everything was so fragmented and the code was duping all over the place.

So agree to disagree I guess.

evil-at-wow commented on src/game/BattleGround/BattleGroundWS.h:
```diff
@@ -101,9 +111,92 @@ enum BG_WS_Events
     WS_EVENT_SPIRITGUIDES_SPAWN   = 2
 };
 
+static const uint32 wsFlagIds[PVP_TEAM_COUNT] = { 
```
It looks like none of the statics in this new block are used in the header, so please move all of this to the source file (for example to the start of the source file, in an anonymous namespace). I know this project isn't always the best example, but it's generally good practice to keep header files clean and minimal, so everything that's not part of "the interface" or "the API" and that can reasonably be kept out of headers, should probably not be in the header.

evil-at-wow commented on src/game/BattleGround/BattleGroundWS.h:
```diff
@@ -115,21 +208,21 @@ class BattleGroundWS : public BattleGround
         void RespawnFlag(Team team, bool captured);
         void RespawnDroppedFlag(Team team);
         uint8 GetFlagState(Team team) { return m_FlagState[GetTeamIndexByTeamId(team)]; }
+        uint32 GetWorldStateUpdateId(uint8 teamIdx);
```
This is not modifiying state, so please make it a const function.

evil-at-wow commented on src/game/BattleGround/BattleGroundWS.h:
```diff
@@ -143,6 +236,10 @@ class BattleGroundWS : public BattleGround
         virtual WorldSafeLocsEntry const* GetClosestGraveYard(Player* player) override;
 
+        // Flag interactions
+        void PickUpFlagFromBase(Player* source, GameObject* target);
```
Unless I missed something, the GameObject parameter isn't used in the implementation. So that parameter can be removed.

Also, you *always* need a valid Player object to be passed to this function, because you're unconditionally dereferencing the parameter in the implementation. That's fine, but that fact is not clear with this API \(*\). A suggestion would be to make it a reference instead, because that makes the intention "a valid object is required" (aka "nullptr is not a valid value") clear even if you only look at the header.

\(*\) I admit that API in the project is often doing it like this though. We should change/improve that IMHO.

evil-at-wow commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,221 +205,156 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
 
-    SendMessageToAll(message_id, type, source);
-    source->RemoveAurasWithInterruptFlags(AURA_INTERRUPT_FLAG_ENTER_PVP_COMBAT);
+    uint8 event = (sBattleGroundMgr.GetGameObjectEventIndex(target->GetGUIDLow())).event1;
```
I think you can remove the outer pair of () here.

evil-at-wow commented on src/game/BattleGround/BattleGroundWS.h:
```diff
@@ -20,11 +20,19 @@
 #define __BATTLEGROUNDWS_H
 
 #include "BattleGround.h"
+#include "Tools/Language.h"
```
This can be moved to the source as well (see remark below about the block of statics). Fewer dependencies in headers is always good :wink:

domgho commented on src/game/BattleGround/BattleGround.cpp:
```diff
@@ -583,6 +583,9 @@ void BattleGround::CastSpellOnTeam(uint32 SpellID, Team teamId)
 void BattleGround::RewardHonorToTeam(uint32 Honor, Team teamId)
 {
+    if (teamId == TEAM_NONE)
+        return;
```
Instead of adding this check in the one required place this now adds unnecessary checks for a team in cases where we already knew that the team is not none and makes it harder harder to reason about the code

BenDol replied:
no team gets no RewardHonorToTeam, given this call is made from EndBattleGround and isn't very intensive it doesn't seem like a reasonable optimization.

domgho replied:
Obviously the performance is negligible in all touched code but why intentionally write more inefficient and harder to understand code than my suggestion

domgho commented on src/game/BattleGround/BattleGroundWS.h:
```diff
@@ -115,21 +124,21 @@ class BattleGroundWS : public BattleGround
-        bool IsHordeFlagPickedUp() const { return !m_flagCarrierHorde.IsEmpty(); }
+        bool IsFlagPickedUp(uint8 teamIdx) const { return !m_FlagCarrier[teamIdx].IsEmpty(); }
+        bool IsAllianceFlagPickedUp() const { return !m_FlagCarrier[TEAM_INDEX_ALLIANCE].IsEmpty(); }
+        bool IsHordeFlagPickedUp() const { return !m_FlagCarrier[TEAM_INDEX_HORDE].IsEmpty(); }
```
The generic version that you added should be called instead of the team versions now

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,221 +288,156 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
-    {
-        m_FlagsTimer[GetOtherTeamIndex(GetTeamIndexByTeamId(source->GetTeam()))] = BG_WS_FLAG_RESPAWN_TIME;
-    }
+        m_FlagsTimer[GetOtherTeamIndex(GetTeamIndexByTeamId(player->GetTeam()))] = BG_WS_FLAG_RESPAWN_TIME;
```
Use `otherTeamIdx`

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,221 +288,156 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
 
+uint32 BattleGroundWS::GetWorldStateUpdateId(uint8 teamIdx) const
+{
+    return teamIdx == TEAM_INDEX_ALLIANCE ? BG_WS_FLAG_UNK_ALLIANCE : BG_WS_FLAG_UNK_HORDE;
```
This should be an array lookup too instead of a function of a branch

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,221 +288,156 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
-        UpdateWorldState(BG_WS_FLAG_UNK_HORDE, 1);
-        source->CastSpell(source, BG_WS_SPELL_WARSONG_FLAG, TRIGGERED_OLD_TRIGGERED);
+        UpdateFlagState(GetOtherTeam(player->GetTeam()), BG_WS_FLAG_STATE_WAIT_RESPAWN);
+        RespawnFlag(player->GetTeam(), false);
```
`player->GetTeam()` should be cached in variable if its used multiple times in a function

Lillecarl replied:
> `player->GetTeam()` should be cached in variable if its used multiple times in a function. Same in other places

What's the benefit of this?

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,221 +288,156 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
+    PvpTeamIndex otherTeamIdx = GetOtherTeamIndex(teamIdx);
 
-    uint8 event = (sBattleGroundMgr.GetGameObjectEventIndex(target_obj->GetGUIDLow())).event1;
+    int32 actionId = -1, messageId = 0, soundId = 0;
```
We use a single line per variable declaration: https://wiki.sei.cmu.edu/confluence/display/c/DCL04-C.+Do+not+declare+more+than+one+variable+per+declaration

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,221 +288,156 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
 
-    // Horde flag on ground(not in base) (returned or picked up again)
-    if (GetFlagState(HORDE) == BG_WS_FLAG_STATE_ON_GROUND && source->IsWithinDistInMap(target_obj, 10))
+    if (actionId > -1)
```
You would get less lines and code complexity by simply calling those three functions in the places where you set the variables

domgho commented on src/game/BattleGround/BattleGround.cpp:
```diff
@@ -583,6 +583,9 @@ void BattleGround::CastSpellOnTeam(uint32 SpellID, Team teamId)
 void BattleGround::RewardHonorToTeam(uint32 Honor, Team teamId)
 {
+    if (teamId == TEAM_NONE)
+        return;
```
Obviously the performance is negligible in all touched code but why intentionally write more inefficient and harder to understand code than my suggestion of moving it to the single needed place in EndBattleGround

BenDol replied:
This is to make sure this call is future proof for other use cases and fits the design of the function which is `Reward Honor To Team` that implies a valid team which `TEAM_NONE` does not fit. This is not something that would be considered harder to understand.

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -150,56 +236,46 @@ void BattleGroundWS::RespawnDroppedFlag(Team team)
-    UpdateFlagState(source->GetTeam(), 1);                  // flag state none
-    UpdateTeamScore(source->GetTeam());
+    UpdateFlagState(player->GetTeam(), 1);                  // flag state none
+    UpdateTeamScore(player->GetTeam());
```
GetTeam() should be cached at start of function

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,218 +293,143 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
-        UpdateWorldState(BG_WS_FLAG_UNK_ALLIANCE, 1);
-        source->CastSpell(source, BG_WS_SPELL_SILVERWING_FLAG, TRIGGERED_OLD_TRIGGERED);
+        actionId = BG_WS_FLAG_ACTION_RETURNED;
+        Team team = player->GetTeam();
```
GetTeam() should be cached at start of function and used in all cases instead of calling the function

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,218 +293,143 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
-    // Alliance flag on ground(not in base) (returned or picked up again from ground!)
-    if (GetFlagState(ALLIANCE) == BG_WS_FLAG_STATE_ON_GROUND && source->IsWithinDistInMap(target_obj, 10))
+    if (actionId > -1)
+        player->RemoveAurasWithInterruptFlags(AURA_INTERRUPT_FLAG_ENTER_PVP_COMBAT);
```
You could still remove this branch by moving this into each section of the previous branch which is more efficient and easier to understand

BenDol replied:
I think since I'm returning what action was executed, I might as well leave this as it is.

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,218 +293,143 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
-            UpdateWorldState(BG_WS_FLAG_UNK_ALLIANCE, uint32(-1));
-        }
+        player->CastSpell(player, SPELL_RECENTLY_DROPPED_FLAG, TRIGGERED_OLD_TRIGGERED);
+        UpdateFlagState(player->GetTeam(), 1);
```
GetTeam() should be cached at start of function and used in all cases where GetTeam is called

domgho commented on src/game/BattleGround/BattleGroundWS.cpp:
```diff
@@ -217,218 +293,143 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
+    m_FlagState[otherTeamIdx] = BG_WS_FLAG_STATE_ON_PLAYER;
+
+    // update world state to show correct flag carrier
+    UpdateFlagState(player->GetTeam(), BG_WS_FLAG_STATE_ON_PLAYER);
```
GetTeam() should be cached at start of function and used in all cases where GetTeam is called

domgho commented:
I think the code is looking better than before and if you take my remaining and still unresolved suggestions into account I would revert my stance of not taking this cleanup since it will probably have no measurable effect on actual live performance. Thanks for your work on this project and your willingness to be open to changes.

berserkingyadis commented:
After testing the flag/capture/point interactions, these are my results:

# Display
- horde flag placement a bit off?
![image](https://user-images.githubusercontent.com/9959527/59563260-c0937100-9037-11e9-8512-16d5eecd21a8.png)
The alliance one appears to be centered well:
![image](https://user-images.githubusercontent.com/9959527/59563265-c5582500-9037-11e9-842c-a922582fe7e7.png)
- top display of who has the flag: ok
- scoring points: ok
- bg happenings in chat, ok, with 1 thing not correct
The ingame output does not appear to be right. This text appears when the flag gets returned automatically
![image](https://user-images.githubusercontent.com/9959527/59563248-993ca400-9037-11e9-9d4b-e8e28a52f554.png)
# Flag interactions
- getting flag from enemy base: ok
- dropping flag by rightclicking: ok
- dropping flag by death: ok
- picking flag back up: ok
- not picking up flag and it gets returned to base automatically: ok
- flags respawining after score: ok
- not being able to score because own flag is gone: ok
# Bg 
- win when 3 flags were captured: ok
- Scoreboard number of flag captures and flag returns are correct: ok

BenDol commented:
@killerwife do we have sniffs for the flag spawn positions?

AnonXS commented:
```
[0] GUID: Full: 0xF11002BE76000B34 Entry: 179830 UInt64: 17370386779814759220
[0] Object Type: GameObject (5)
[0] Update Flags: Unknown1, LowGuid, StationaryObject (88)
[0] Stationary Position: X: 1540.423 Y: 1481.325 Z: 351.8284, O: 3.089183

[3] GUID: Full: 0xF11002BE77000B35 Entry: 179831 UInt64: 17370386779831536437
[3] Object Type: GameObject (5)
[3] Update Flags: Unknown1, LowGuid, StationaryObject (88)
[3] Stationary Position: X: 916.5073 Y: 1433.826 Z: 346.3796, O: 0.01745246
```

BenDol commented:
@berserkingyadis fixed the positioning here https://github.com/cmangos/tbc-db/pull/655/files and also resolved the message issues in the latest commit. As for the spells dropping the flag, I have the feeling those spells aren't supposed to drop the flag, but could be wrong.

berserkingyadis commented:
These are the new Flag positions. Alliance flag looks good as ever and Horde flag still appears not to be in the middle of these skulls but can't argue with sniffs, I guess.
![flags](https://user-images.githubusercontent.com/9959527/59582592-0fddae00-90d9-11e9-8a2e-727257c645ad.jpg)
Message issues are now fixed. The new strings are:
"The Horde flag is now placed at its base."
"The Alliance flag is now placed at its base."

Regarding the spells dropping the flag, can we get some BG pros in here? Alternatively just feedback how it should work with these spells (and other spells).

AnonXS commented:
The Spells you mentioned (SS/CoS/Blink) should not drop flag afaik.

killerwife commented:
Closed by https://github.com/cmangos/mangos-tbc/commit/f4fd912e2f3253bd67625361224faa0fa31a1fc3
Thx for the rework.

henhouse commented:
Hey @BenDol, sorry to bump a closed PR. Was looking over this and saw that the debuff logic is apparently missing from cMaNGOS entirely. Did you intentionally leave it out in this rework, or were unaware of it by chance?

Spells are `46392` and `46393`.

See comment here for explanation on how they work: https://www.wowhead.com/spell=46393/brutal-assault#comments:id=290484
